src/node_http2.cc

```diff
@@ -2379,6 +2416,44 @@ void Http2Stream::RefreshState(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+void Http2Session::AltSvc(int32_t id,
+                          uint8_t* origin,
+                          size_t origin_len,
+                          uint8_t* value,
+                          size_t value_len) {
+  Http2Scope h2scope(this);
+  nghttp2_submit_altsvc(session_, NGHTTP2_FLAG_NONE, id,
+                        origin, origin_len, value, value_len);
+}
+
+// Submits an AltSvc frame to the sent to the connected peer.
+void Http2Session::AltSvc(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  Http2Session* session;
+  ASSIGN_OR_RETURN_UNWRAP(&session, args.Holder());
+
+  int32_t id = args[0]->Int32Value(env->context()).ToChecked();
+
+  // origin and value are both required to be ASCII, handle them as such.
```

Comment:
Is this requirement enforced for the value string? If not then we should, to combat the HTTP header encoding debacle that only recently got rectified. In fact IMO we should verify that it only consists of HTTP tokens, similar to the HTTP header checks. 

Origin is fine because it’s guaranteed to be ASCII by the URL parser.

Reply:
The value is not limited to `token`. It will contain a sequence of `uri-host` constructs (`host:port`) within quoted strings. At the C++ level, we write these out as one-byte strings so there is at least some level of enforcement there. And on the *receiving* side, these are treated as one byte strings so the value will be received correctly. Anyone using extended characters will just see those get corrupted. Because of the performance implications and the fact that this is a non-critical API, I don't think we need to validate every character.... although, it would be good to extend the documentation on this a bit more -- as you said, not everyone likes to read RFCs :-)

Reply:
The additional documentation is always a plus. 

> Anyone using extended characters will just see those get corrupted.

Well that’s certainly not the best of outcomes. If this is truly a non-critical API, I think we should ensure correctness above performance. Good call on quotation marks though, and in that case I think simply using a `/[\x00-\x7f]/` check should be enough.

Reply:
There may also be some security implications  involved, when `'\uff20'` is treated identically to `'\u0020' by the V8 API.

Reply:
Ok, added a check